_posts/2019-05-10-totw-45.md

```diff
@@ -0,0 +1,65 @@
+---
+title: "Tip of the Week #45: Avoid Flags, Especially in Library Code"
+layout: tips
+sidenav: side-nav-tips.html
+published: true
+permalink: tips/45
+type: markdown
+order: "045"
+---
+
+Originally posted as TotW #45 on June 3, 2013
+
+*by [Titus Winters](mailto:titus@google.com)
+
+*"What I really want is the behavior of my code to be controlled by a global
+variable that cannot be statically predicted, whose usage is incompletely
+logged, and which can only be removed from my code with great difficulty." --
+Nobody, Ever*
+
+The common use of flags in production code, especially within libraries, is
+a mistake. Don’t use flags unless it is truly necessary. There, we said it.
+
+Flags are global variables, only worse: you can’t know the value of that
+variable by reading the code. A flag may not only be set at startup, but could
+potentially be changed later through most of our common HTTP interfaces. If you
+run a server in your binary, there is usually no guarantee that the value of
+your flag will remain unchanged from cycle to cycle, nor any notification if
+it were to change, nor any means for looking for such a change.
+
+Even if you ignore the network-based control "feature", it is often hard (and
+sometimes impossible) to reason about flag values, since flags, by design, may
+be overridden at startup. This uncertainty would be OK if we reliably logged
+binary invocations in production, but we don’t: many production flags are not
```

Comment:
This is all about internal production logging, should be dropped or restructured.

"If your production environment logs the invocation of every binary directly, and stores those logs, great. 
Most environments don't work that way."